src/algebra/algebra/unitization.lean

```diff
@@ -0,0 +1,505 @@
+
+variables (R A)
+
+/-- The canonical inclusion of rings `R →+* unitization R A`. -/
+@[simps apply]
+def inl_ring_hom [semiring R] [non_unital_semiring A] [module R A] : R →+* unitization R A :=
+{ to_fun := inl,
+  map_one' := inl_one A,
+  map_mul' := inl_mul A,
+  map_zero' := inl_zero A,
+  map_add' := inl_add A }
+
+end mul
+
+section algebra
+variables (S R A : Type*)
+[comm_semiring S] [comm_semiring R] [non_unital_semiring A]
+[module R A] [is_scalar_tower R A A] [smul_comm_class R A A]
+[algebra S R] [distrib_mul_action S A] [is_scalar_tower S R A]
+
+instance algebra : algebra S (unitization R A) :=
+{ commutes' := λ r x,
+  begin
+    induction x using unitization.ind,
+    simp only [mul_add, add_mul, ring_hom.to_fun_eq_coe, ring_hom.coe_comp, function.comp_app,
+      inl_ring_hom_apply, inl_mul_inl],
+    rw [inl_mul_inr, inr_mul_inl, mul_comm]
+  end,
+  smul_def' := λ s x,
+  begin
+    induction x using unitization.ind,
+    simp only [mul_add, smul_add, ring_hom.to_fun_eq_coe, ring_hom.coe_comp, function.comp_app,
+      inl_ring_hom_apply, algebra.algebra_map_eq_smul_one],
+    rw [inl_mul_inl, inl_mul_inr, smul_one_mul, inl_smul, inr_smul, smul_one_smul]
+  end,
+  ..(unitization.inl_ring_hom R A).comp (algebra_map S R) }
+
+lemma algebra_map_eq_inl_comp : ⇑(algebra_map S (unitization R A)) = inl ∘ algebra_map S R := rfl
+lemma algebra_map_eq_inl_ring_hom_comp :
+  algebra_map S (unitization R A) = (inl_ring_hom R A).comp (algebra_map S R) := rfl
+lemma algebra_map_eq_inl : ⇑(algebra_map R (unitization R A)) = inl := rfl
+lemma algebra_map_eq_inl_hom : algebra_map R (unitization R A) = inl_ring_hom R A := rfl
+
+/-- The canonical `R`-algebra projection `unitization R A → R`. -/
+@[simps]
+def fst_hom : unitization R A →ₐ[R] R :=
+{ to_fun := fst,
+  map_one' := fst_one,
+  map_mul' := fst_mul,
+  map_zero' := fst_zero,
+  map_add' := fst_add,
+  commutes' := fst_inl A }
+
+end algebra
+
+section coe
+variables {R A : Type*}
+
+instance [has_zero R] : has_coe_t A (unitization R A) := { coe := inr }
+
+@[simp]
+lemma coe_apply [has_zero R] (a : A) : (a : unitization R A) = inr a := rfl
+
+lemma coe_injective [has_zero R] : function.injective (coe : A → unitization R A) :=
+inr_injective
+
+/-- The coercion from a non-unital `R`-algebra `A` to its unitization `algebra.unitization R A`
+realized as a non-unital algebra homomorphism. -/
+@[simps]
+def coe_non_unital_alg_hom [comm_semiring R] [non_unital_semiring A] [module R A] :
+  non_unital_alg_hom R A (unitization R A) :=
+{ to_fun := coe,
+  map_smul' := by simp,
+  map_zero' := by simp,
+  map_add' := by simp,
+  map_mul' := by simp }
+
+end coe
+
+section alg_hom
+
+variables
+{S R A : Type*}
+[comm_semiring S] [comm_semiring R] [non_unital_semiring A]
+[module R A] [smul_comm_class R A A] [is_scalar_tower R A A]
+{B : Type*} [ring B] [algebra S B]
+[algebra S R] [distrib_mul_action S A] [is_scalar_tower S R A]
+{C : Type*} [ring C] [algebra R C]
```

Comment:
I think normally we indent these, or split them across separate `variables` lines